src/driver/driver.cpp

```diff
@@ -61,17 +61,23 @@ DriverStatus EvolutionDriver::Execute() {
   while (tm.KeepGoing()) {
     if (Globals::my_rank == 0) OutputCycleDiagnostics();
 
+    pmesh->PreStepUserWorkInLoop(pmesh, pinput, tm);
+    pmesh->PreStepUserDiagnosticsInLoop(pmesh, pinput, tm);
+
     TaskListStatus status = Step();
     if (status != TaskListStatus::complete) {
       std::cerr << "Step failed to complete all tasks." << std::endl;
       return DriverStatus::failed;
     }
-    // pmesh->UserWorkInLoop();
+
     tm.ncycle++;
     tm.time += tm.dt;
     pmesh->mbcnt += pmesh->nbtotal;
     pmesh->step_since_lb++;
 
+    pmesh->PostStepUserWorkInLoop(pmesh, pinput, tm);
+    pmesh->PostStepUserDiagnosticsInLoop(pmesh, pinput, tm);
+
```

Comment:
Are the function call after 
```
    tm.ncycle++;
    tm.time += tm.dt;
```
on purpose?
Conceptually, I would put them before because in case I add have some userwork that should be done in a specific cycle/time, I could filter for the same cycle/time in both pre and post functions, e.g.,
```
PreStepUserDiagnosticsInLoop (pmesh, pinput, tm) {
  if (tm.ncycle % 10 == 0) {
    my_magic_diagnostic_before_step();
  }
}
```
and PostStep correspondingly.

Otherwise, I'd need `ncycle` and `ncycle+1`, which I personally find less intuitive.

Reply:
I think I did this because the diagnostic and work is occurring after the step has happened, so the time the problem state has reached is "effectively" at the updated time.  But, one could use the `PreStep` stuff on the next cycle to do that provided the problem hasn't terminated.

Reply:
@pgrete : I have moved those calls to before the update to time and cycle.